File: sqlfluffrs/sqlfluffrs_types/src/identity.rs
//! Process-global unique identifiers for tokens.
//!
//! Replaces per-token `Uuid::new_v4()`. A token's `uuid` is only ever used as
//! an opaque identity key (compared by equality and used as a map key in the
//! fix engine); it is never parsed back as an RFC-4122 UUID, so a monotonic
//! counter is sufficient -- and far cheaper than drawing 16 bytes of CSPRNG
//! entropy per token.
//!
//! The high bits carry a source tag so these ids stay disjoint from ids minted
//! on the Python side (which uses `PYTHON_TAG = 1 << 120`). `PyToken.uuid`
//! surfaces this `u128` to Python verbatim, where Python-created segments and
//! Rust-origin tokens share one id space inside the fix engine. Without the
//! tag, both counters would start at 1 and a Rust token would collide with an
//! unrelated Python segment.
//!
//! NOTE: this is the token *identity* uuid only. Templated `block_uuid`s remain
//! genuine `Uuid`s -- they are round-tripped through `Uuid::parse_str` at the
//! Python boundary and must keep RFC-4122 structure.

use std::sync::atomic::{AtomicU64, Ordering};

/// Source tag OR'd into the high bits of every Rust-minted id. Must stay
/// distinct from the Python tag (`1 << 120`). `2 << 120` sets bit 121, leaving
/// the low 120 bits for the counter; both fit within a `u128`.
const RUST_TAG: u128 = 2u128 << 120;

static COUNTER: AtomicU64 = AtomicU64::new(1);

/// Return a process-unique, source-tagged token identifier.
///
/// `fetch_add` with `Relaxed` ordering is correct here: we need atomic
/// uniqueness, not any happens-before relationship with other memory, so the
/// cheapest ordering suffices.
pub fn next_id() -> u128 {
    RUST_TAG | (COUNTER.fetch_add(1, Ordering::Relaxed) as u128)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Tag must match Python's convention and never overlap its space.
    const PYTHON_TAG: u128 = 1u128 << 120;

    #[test]
    fn ids_are_tagged_unique_and_disjoint_from_python() {
        let a = next_id();
        let b = next_id();
        assert_ne!(a, b, "ids must be unique");
        assert_eq!(a >> 120, 0b10, "rust tag (bit 121) must be set");
        // A Rust id can never equal a Python id: their tag bits differ.
        assert_ne!(a & !(RUST_TAG | PYTHON_TAG), a, "counter occupies low bits");
        assert_eq!(RUST_TAG & PYTHON_TAG, 0, "tag spaces must not overlap");
    }
}

File: src/sqlfluff/core/helpers/identity.py
"""Process-global unique identifiers for segments.

Replaces per-instance ``uuid4()`` generation. A monotonic counter is
sufficient because segment ``uuid`` values are only ever used as opaque,
in-memory identity keys (dict keys and ``==`` in the fix engine). They never
need cryptographic unpredictability or cross-process uniqueness, and are never
parsed back as RFC-4122 UUIDs.

The high bits carry a source tag so these ids stay disjoint from token ids
minted on the Rust side (see the ``identity`` module in ``sqlfluffrs_types``,
which uses ``RUST_TAG = 2 << 120``). Both runtimes count from 1, so without the
tag a Python-created segment and a Rust-origin token would share low integer
ids -- and the fix engine would then treat two unrelated segments as the same
anchor. This is not hypothetical: ``PyToken.uuid`` surfaces the Rust ``u128``
to Python verbatim, so the two id spaces genuinely meet inside one tree.
"""

from __future__ import annotations

import itertools

# Source tag OR'd into the high bits of every Python-minted id. Must stay
# distinct from the Rust tag (``2 << 120``). Fits within 128 bits alongside any
# realistic process-lifetime counter value (which is always < ``2 ** 120``).
_PYTHON_TAG = 1 << 120

_counter = itertools.count(1)


def get_next_id() -> int:
    """Return a process-unique, source-tagged identifier.

    ``next()`` on an ``itertools.count`` is atomic under the CPython GIL -- its
    C-level ``__next__`` never releases the GIL mid-increment -- so this is
    thread-safe without an explicit lock. Avoiding the lock is the point: it
    keeps the per-segment cost to a single counter increment rather than
    re-introducing the lock acquire/release that motivated dropping ``uuid4``.
    """
    return _PYTHON_TAG | next(_counter)
